libs/pilight/protocols/433.92/dogcollar.c

```diff
@@ -0,0 +1,184 @@
+#include "dogcollar.h"
+
+#define PULSE_MULTIPLIER	3
+#define MIN_PULSE_LENGTH	408
+#define MAX_PULSE_LENGTH	398
+#define AVG_PULSE_LENGTH	403
+#define RAW_LENGTH				50
+
+
+static int validate(void) {
+	if(dogcollar->rawlen == RAW_LENGTH) {
+		if(dogcollar->raw[dogcollar->rawlen-1] >= (MIN_PULSE_LENGTH*PULSE_DIV) &&
+		   dogcollar->raw[dogcollar->rawlen-1] <= (MAX_PULSE_LENGTH*PULSE_DIV)) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static void createMessage(int id, int function) {
+        dogcollar->message = json_mkobject();
+        json_append_member(dogcollar->message, "id", json_mknumber(id,0));
+        json_append_member(dogcollar->message, "function", json_mknumber(function,0));
+}
+
+
+static void createLow(int s, int e) {
+	int i;
+
+	for(i=s;i<=e;i+=2) {
+		dogcollar->raw[i]=(AVG_PULSE_LENGTH);
+		dogcollar->raw[i+1]=(PULSE_MULTIPLIER*AVG_PULSE_LENGTH);
+	}
+
+}
+
+static void createHigh(int s, int e) {
+	int i;
+
+	for(i=s;i<=e;i+=2) {
+		dogcollar->raw[i]=(PULSE_MULTIPLIER*AVG_PULSE_LENGTH);
+		dogcollar->raw[i+1]=(AVG_PULSE_LENGTH);
+	}
+}
+static void clearCode(void) {
+        createLow(0,48);
+        createHigh(36,37);
+}
+
+static void createId(int id) {
+        int binary[255];
+        int length = 0;
+        int i=0, x=0;
+
+        length = decToBinRev(id, binary);
+	for(i=0;i<=length;i++) {
+                if(binary[i]==1) {
+                        x=i*2;
+                        createHigh(x,x+1);
+                }
+        }
+}
+
+static void createFooter(void) {
+        dogcollar->raw[48]=(2*AVG_PULSE_LENGTH);
+        dogcollar->raw[49]=(PULSE_DIV*AVG_PULSE_LENGTH);
+}
+
+static void createFunction(int function) {
+	if(function == 1)
+     	createHigh(46,47);  //light
+    if(function == 2)
+        createHigh(44,45);  //beep
+    if(function == 3)
+        createHigh(42,43);  //vibrate low
+    if(function == 4)
+        createHigh(40,41);  //vibrate high
+}
+
+
+static int createCode(struct JsonNode *code) {
+	int id = -1;
+	int function = -1;
+	double itmp = 0;
+
+	if(json_find_number(code, "id", &itmp) == 0)
+		id = (int)round(itmp);
+	if(json_find_number(code, "function", &itmp) == 0)
+		function = (int)round(itmp);
+
+	if(id == -1 || function == -1) {
+		logprintf(LOG_ERR, "dogcollar: insufficient number of arguments");
+		return EXIT_FAILURE;
+	} else if(id > 29999 || id < 1) {
+		logprintf(LOG_ERR, "dogcollar: invalid id range");
+		return EXIT_FAILURE;
+	} else if(function > 4 || function < 1) {
+		logprintf(LOG_ERR, "dogcollar: invalid function");
+		return EXIT_FAILURE;
+	} else {
+        createMessage(id, function);
+		clearCode();
+        createId(id);
+        createFunction(function);
+        createFooter();
+		dogcollar->rawlen = RAW_LENGTH;
+	}
+	return EXIT_SUCCESS;
+}
+
+static void printHelp(void) {
+        printf("\t -i --id=id\t\tcontrol a device with this id (1-29999)\n");
+        printf("\t -f --function=function\t\t (1=light, 2=beep, 3=vibrate-low, 4=vibrate-high)\n");
```

Comment:
Hi CurlyMoo,
There are no controls for the GUI that would support this change, hence why I had to go with a numeric value. If you can point me in the direction otherwise then I will be happy to re-design this protocol structure as you see fit. The only options of type JSON_STRING are states which have a text value (upon button press) returned of either "down" or "up"